magenta/js/music_rnn/src/model.ts

```
@@ -0,0 +1,178 @@
+
+const DEFAULT_MIN_NOTE = 48;
+const DEFAULT_MAX_NOTE = 83;
+
+const CELL_FORMAT = "rnn/multi_rnn_cell/cell_%d/basic_lstm_cell/";
+
+/**
+ * Main MusicVAE model class.
+ *
+ * A MusicVAE is a variational autoencoder made up of an `Encoder` and
+ * `Decoder`, along with a `DataConverter` for converting between `Tensor`
+ * and `NoteSequence` objects for input and output.
+ *
+ * Exposes methods for interpolation and sampling of musical sequences.
+ */
+export class MelodyRnn {
+  checkpointURL: string;
+
+  lstmCells: dl.LSTMCellFunc[];
+  lstmFcB: dl.Tensor1D;
+  lstmFcW: dl.Tensor2D;
+  forgetBias: dl.Scalar;
+  biasShapes: number[];
+  rawVars: {[varName: string]: dl.Tensor};  // Store for disposal.
+
+  initialized: boolean;
+
+  /**
+   * `MusicVAE` constructor.
+   *
+   * @param checkpointURL Path to the checkpoint directory.
+   * @param dataConverter A `DataConverter` object to use for converting between
+   * `NoteSequence` and `Tensor` objects. If not provided, a `converter.json`
+   * file must exist within the checkpoint directory specifying the type and
+   * args for the correct `DataConverter`.
+   */
+  constructor(checkpointURL=CHECKPOINT_URL) {
+    this.checkpointURL = checkpointURL;
+    this.initialized = false;
+    this.rawVars = {};
+    this.biasShapes = [];
+    this.lstmCells = [];
+  }
+
+  /**
+   * Loads variables from the checkpoint and instantiates the `Encoder` and
+   * `Decoder`.
+   */
+  async initialize() {
+    this.dispose();
+
+    const reader = new dl.CheckpointLoader(this.checkpointURL);
+    const vars = await reader.getAllVariables();
+
+    this.forgetBias = dl.scalar(1.0);
+
+    this.lstmCells.length = 0;
+    this.biasShapes.length = 0;
+    let l = 0;
+    while (true) {
+      const cellPrefix = CELL_FORMAT.replace('%d', l.toString());
+      if (!(cellPrefix + 'kernel' in vars)) {
+          break;
+      }
+      this.lstmCells.push(
```

Can you add a check and a TODO for attention? I think we will silently ignore it now, giving bad results.

---

Done

---

Could you just add something to see if the attention variable is there and throw an error if it is?